Add new Analytics provisioning 'AccountCreate' component.

adamsilverstein opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #1211

## Relevant technical choices

<!-- Please describe your changes. -->

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,148 @@
+import PropertyField from './property-field';
+import ProfileField from './profile-field';
+
+// import Data from 'googlesitekit-data';
```
The commented lines implement the data connection and can be uncommenting once that part is merged.

adamsilverstein replied:
Uncommented.

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,163 @@
+
+const AccountCreate = () => {
+	// const { createAccount } = dispatch( STORE_NAME );
+	// const isDoingCreateAccount = select( STORE_NAME ).isDoingCreateAccount();
```
@aaemnnosttv I may need some help using this store element to control the 'submitting' state here. Previously I tracked this in component state, however I'm trying to use the store instead.

I've tried using `withSelect`, however the select element passed to the callback there doesn't have access to the correct store. Using `select` from Data directly does work, but only runs once (eg the component doesn't update when the store value changes).

I tried something like this:
```js
const isDoingCreateAccount = useSelect(
		( select ) => {
			return select( STORE_NAME ).isDoingCreateAccount();
		},
		[]
	);
```

adamsilverstein replied:
Right! Thanks for clarifying.

adamsilverstein replied:
Added this and switched to `useDispatch` as well (commented out for now).

aaemnnosttv commented on assets/js/modules/analytics/common/create-account-field.js:
```diff
@@ -0,0 +1,60 @@
+			label={ label }
+			name={ name }
+			onChange={ ( e ) => {
+				validationIssues[ name ] = '' === e.target.value;
```
This is mutating the prop which could cause problems if that in turn changes the value outside of the component by reference.

Better to just merge it in before/while setting:
```js
setValidationIssues( {
    ... validationIssues,
    [ name ] : '' === e.target.value,
} );
```

adamsilverstein replied:
Great suggestion, thank you!

adamsilverstein replied:
Updated in https://github.com/google/site-kit-wp/pull/1389/commits/d445f69708dc5f8634749d78613eb4680eb3b509.

adamsilverstein commented on assets/js/modules/analytics/common/countries.js:
```diff
@@ -0,0 +1,3042 @@
+export const countries = {
```
I considered an `npm` package as well as using WordPress's Timezone data. 

The goal here is to provide a timezone selection experience that closely mirrors what you get when going thru the normal Analytics setup flow (with the added benefit of your WordPress timezone being pre-selected).  

The timezones available in the API aren't documented and WordPress at least provides some non standard options, I wasn't sure any package would have the correct set of options so am trying to match what is available in the regular setup flow.

The data I am including here is the data from an API call that is used during the regular (web based) Analytics setup flow. You can see the call to this 'countries' API endpoint if you go thru the flow to create a new account. The API is undocumented and requires authentication so I'm capturing the returned data to use here.

adamsilverstein commented on includes/Modules/Analytics.php:
```diff
@@ -133,6 +133,19 @@ function() {
 		);
+
+		// Add the timezone and errorcode data for users provisioning new accounts.
+		if ( ! $this->is_connected() ) {
```
good point, I've mainly been testing the no account yet setup flow but this absolutely should work to provision new accounts for existing users. I guess in this case if the user selects 'New Account' we would display the new component inline and need to offer a cancel option.

adamsilverstein replied:
Always add the data in https://github.com/google/site-kit-wp/pull/1389/commits/90cd076d4fab6b35b2a86483c3b66798c545ec91.

I will address the 'Create new account from settings page in a follow up Issue/PR.

adamsilverstein commented on includes/Modules/Analytics.php:
```diff
@@ -133,6 +133,19 @@ function() {
+				function( $admin_data ) {
+					$selected_zone           = get_option( 'timezone_string' );
+					$admin_data['timezone']  = $selected_zone;
+					$admin_data['errorcode'] = $this->context->input()->filter( INPUT_GET, 'error_code', FILTER_SANITIZE_STRING );
```
nope, that makes sense. I can switch it to use the client side calls.

adamsilverstein replied:
Moved in https://github.com/google/site-kit-wp/pull/1389/commits/1c3924981d453ad6286969a474d56af968f21e01.

adamsilverstein commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,137 @@
+	Select,
+	Option,
+} from '../../../material-components';
+import { countries } from './countries';
+const { default: { country: allCountries } } = countries;
```
As explained above I want to ensure the data and behavior of the timzone selector matches that on the Analytics web setup flow. For that reason I have chosen to use the same data it uses.

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,163 @@
+						<h2>
+							{ __( 'Create new Analytics account', 'google-site-kit' ) }
+						</h2>
+						{
+							error &&
+							<div className="error">
+								<p>
+									{ error }
+								</p>
+							</div>
+						}
```
great, I will for that. This was a temporary solution that needs improvement.

adamsilverstein replied:
Updated in https://github.com/google/site-kit-wp/pull/1389/commits/d445f69708dc5f8634749d78613eb4680eb3b509.

adamsilverstein replied:
Do you mean using `GenericError` or some other component? (/assets/js/components/notifications/generic-error.js)

adamsilverstein replied:
@aaemnnosttv ^^

felixarntz replied:
There is an `ErrorNotice` component specific to Analytics, in `js/modules/analytics/common`. That automatically uses `getError`, so based on my comments above I think you wouldn't even need to call `getError` at all in this component, and rather just render the `ErrorNotice` component.

adamsilverstein replied:
Since the callback handler is displayed on analytics in an iframe, it makes more sense to show any error messaging there. I am removing the error handling from this component.

adamsilverstein replied:
removed in https://github.com/google/site-kit-wp/pull/1389/commits/d007a17391c76d7f19b81caa09f55fe808b483c9

adamsilverstein commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,137 @@
+	// Track whether the selected country has more than one timezone.
+	let multiTimezone = false;
+
+	timezoneSelectorCache = (
```
The construction of the dropdown is complex, and we need to iterate thru the countries/timezones to find the currently selected country/timezone. Without the caching, typing into other inputs on the page was very slow. (to test, remove the caching).

adamsilverstein replied:
I switched the caching mechanism to leverage `useMemo` in https://github.com/google/site-kit-wp/pull/1389/commits/0c4c796282d72590a1a73d81646dd8a944b7a020.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+	if ( accountTicketTermsOfServiceURL ) {
+		location = accountTicketTermsOfServiceURL;
+	}
+	*/
```
#1311 is merged now :) Also would be good to refresh this against the latest `feature/ga-provisioning` branch.

adamsilverstein replied:
Refreshed and enabled data handling.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+	*/
+
+	const isDoingCreateAccount = false;
+	const { createAccount } = () => {};
```
I guess this was temporary as well :)

adamsilverstein replied:
Yep, removed.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+
+	const isDoingCreateAccount = false;
+	const { createAccount } = () => {};
+	const { siteName, siteURL, timezone: tz } = global.googlesitekit.admin;
```
We should avoid access any globals. All data should come from a datastore - if it's not there yet, we can add it. For each of the above three, the `core/site` store would be a good place.

adamsilverstein replied:
Can you please point me to an example where we are adding these to the store? Should the store initialization use the global (or if not, how do we pass the value from PHP)?

adamsilverstein replied:
I figured it out and added in https://github.com/google/site-kit-wp/pull/1389/commits/4de6bc6e7506c466f74f3dae27076daace5db6a1.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+	} );
+
+	const [ accountName, setAccountName ] = useState( siteName );
+	const [ propertyName, setPropertyName ] = useState( siteURL );
```
Per the ACs (and current PHP behavior) this should only be the host, not the full URL.

adamsilverstein replied:
Good point. Updated in https://github.com/google/site-kit-wp/pull/1389/commits/4060a8e591a4536115641be8a16500408b8aea54.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+			errorMessage = __( 'Unknown service error.', 'google-site-kit' );
+			break;
+	}
+	const [ error, setError ] = useState( errorMessage );
```
This shouldn't be necessary. There's a `getError()` selector in the `modules/analytics` datastore that will hold the API error if the `createAccount` action failed.

Potentially we should incorporate the logic to parse the `error_code` query arg to an error (should be an object with `code` _and_ `message`) into a resolver for `getError`, so that this is handled in the datastore, rather than in the component.

adamsilverstein replied:
Currently handling the error at the proxy so I have removed this section of the code.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+			if ( err ) {
+				setError( err.message ? err.message : __( 'Unknown error.', 'google-site-kit' ) );
+			}
+		} );
```
No need for error handling here based on the above comments, when we start using `getError` it will automatically be displayed. This code here can just become `await createAccount( ... )` (via `useDispatch`) - and afterwards we can do the redirect if the URL (via `useSelect` on `getAccountTicketTermsOfServiceURL`) is not empty.

adamsilverstein replied:
removed

adamsilverstein replied:
not seeing the error handling, let's discuss further.

adamsilverstein replied:
Ok, so I guess `ErrorNotice` is missing, should I go ahead and add that into the component? or should it be on the container component? I'll try placing it here in the component.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+	} );
+
+	// Disable the submit button if there are validation errors, and while submission is in progress.
+	const buttonDisabled = validationIssues.accountName || validationIssues.propertyName || validationIssues.profileName || validationIssues.timezone || isDoingCreateAccount;
```
The first four conditions could be simplified to `Object.values( validationIssues ).some( ( check ) => check )`.

adamsilverstein replied:
Nice! updated in https://github.com/google/site-kit-wp/pull/1389/commits/338c9e51b2a95fab210c239583f7d1fd32dee7a2.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,195 @@
+						<div className="mdc-layout-grid">
+							{
+								isDoingCreateAccount
+									? <ProgressBar />
```
This creates unnecessary indentation. Let's use an early return instead, further above `if ( isDoingCreateAccount ) { return <ProgressBar />; }`.

Also let's add `useState` for an `[ isNavigating, setIsNavigating ]` as we discussed the other day, to keep the loading state going until redirected to GA TOS.

adamsilverstein replied:
Added...  part in https://github.com/google/site-kit-wp/pull/1389/commits/d5ff025785f3a7cc1a12566392b835c6c31d5f01.

felixarntz commented on assets/js/modules/analytics/common/create-account-field.js:
```diff
@@ -0,0 +1,62 @@
+				setValidationIssues( {
+					...validationIssues,
+					[ name ]: '' === e.target.value,
+				} );
```
Setting the validation issues from here results in duplicated code and can be simplified. We can rely on the `useEffect` hook in the main component to re-check validation whenever one of these four values changes. No need to re-handle it here. It will also nicely reduce the amount of props necessary.

adamsilverstein replied:
Great suggestion! Switched to `useEffect` in https://github.com/google/site-kit-wp/pull/1389/commits/f8296c5a2fd1f65e1721662100349bef99872c96.

felixarntz commented on assets/js/modules/analytics/common/create-account-field.js:
```diff
@@ -0,0 +1,62 @@
+		<TextField
+			className={ classnames(
+				'mdc-text-field',
+				{ 'mdc-text-field--error': validationIssues[ name ] }
```
Similarly, let's simplify this prop to be more targeted, something like `hasError`. The `validationIssues` part can then remain internal to the main component.

adamsilverstein replied:
update with `hasError` in https://github.com/google/site-kit-wp/pull/1389/commits/527201e0a64fca7b4eae50a402147fe29fd0417f.

felixarntz commented on assets/js/modules/analytics/common/create-account-field.js:
```diff
@@ -0,0 +1,62 @@
+	validationIssues,
+	setValidationIssues,
+	objectValue,
+	objectSetter,
```
These two names sound a bit odd. Why not just `value` and `setValue`?

adamsilverstein replied:
Naming is hard, I have no idea why I came up with those names. Your suggestion is much better, changed in https://github.com/google/site-kit-wp/pull/1389/commits/b12378a778fb80f349d5e670d860a9dc71dbc808.

felixarntz commented on assets/js/modules/analytics/common/index.js:
```diff
@@ -16,7 +16,7 @@
  */
 
-export { default as AccountCreate } from './account-create-legacy';
+export { default as AccountCreate } from './account-create';
```
As discussed the other day, we will need to include both components here, and use the legacy one if the site is not using the proxy. Let's add a selector `usingProxy` to the `core/site` store for that purpose, and set it based on whether the `_googlesitekitBaseData.proxySetupURL` value exists or not.

adamsilverstein replied:
will do.

adamsilverstein replied:
Added this in https://github.com/google/site-kit-wp/pull/1389/commits/4fd7501de62b7bb620282230bfcf14268ef3177a.

felixarntz commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,140 @@
+		let foundTimezone = false;
+		const response = (
+			<div >
+				<span style={ { minWidth: '240px', margin: '0 5px 0 0' } /*todo: move to css */ } >
```
Let's do this and eliminate the TODO :)

adamsilverstein replied:
👍

adamsilverstein replied:
Moved styles to Sass and improved in https://github.com/google/site-kit-wp/pull/1389/commits/0a9ee72421bc52d6f04ab325d5e384c265664f5c.

felixarntz commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,140 @@
+
+const { default: { country: allCountries } } = countries;
+
+const TimezoneSelect = ( { timezone, setTimezone, validationIssues, setValidationIssues } ) => {
```
Same concern as above here, let's remove the `setValidationIssues` prop and use a `hasError` prop instead of the more complex `validationIssues` one.

adamsilverstein replied:
All fixed in https://github.com/google/site-kit-wp/pull/1389/commits/527201e0a64fca7b4eae50a402147fe29fd0417f

felixarntz commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,140 @@
+								} ) }
+					</Select>
+				</span>
+				<span style={ { minWidth: '240px', margin: '5px 5px 0 0' } /*todo: move to css */ } >
```
Same here, let's do it.

adamsilverstein replied:
Done.

felixarntz commented on includes/Modules/Analytics.php:
```diff
@@ -134,6 +134,16 @@ function() {
+				$admin_data['timezone'] = $selected_zone;
+				return $admin_data;
+			}
+		);
```
Let's remove this and instead expose it on `_googlesitekitBaseData`, so that the `core/site` store can then expose it via a `getTimezone()` selector. Also, this should probably happen right in the main function in the `Assets` class, this value is nothing specific to Analytics.

adamsilverstein replied:
👍 I'll try changing this.

adamsilverstein replied:
Moved data to datastore in https://github.com/google/site-kit-wp/pull/1389/commits/4de6bc6e7506c466f74f3dae27076daace5db6a1 and https://github.com/google/site-kit-wp/pull/1389/commits/63e53c28b307b1ead697a8c36b575bf50b622476

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,166 @@
+			timezone,
+		} ).then( () => {
+			// Log error message?
+			setIsNavigating( false );
```
I guess we do need some error handling here in case the request fails. I will re-add that.

adamsilverstein commented:
Pushed updates for most of the issues you raised @felixarntz and left one question. I'll work on the outstanding items next. All changes so far are merged to https://github.com/google/site-kit-wp/pull/1343 for testing.

adamsilverstein commented:
@felixarntz I have addressed all your feedback and this is ready for another review. One question I have is around error capture (for example if the server fails to respond or throws a server error) - in my testing these are not currently handled, I'm not sure the best approach to add this handling, let's discuss further. (See https://github.com/google/site-kit-wp/pull/1389#discussion_r417589271)

adamsilverstein commented on includes/Modules/Analytics.php:
```diff
@@ -521,7 +521,7 @@ protected function handle_provisioning_callback() {
 		// Retrieve the internal web property id.
 		try {
-			$web_property = $this->get_service( 'analytics' )->management_webproperties->get( $account_id, $property_id );
+			$web_property = $this->get_service( 'analytics' )->management_webproperties->get( $account_id, $web_property_id );
```
fixes a php fatal: the variable was previously renamed and this instance missed.

adamsilverstein commented on includes/Modules/Analytics/Proxy_Provisioning.php:
```diff
@@ -33,6 +33,6 @@ class Proxy_Provisioning extends Google_Service_Resource {
 		$params = array( 'postBody' => $post_body );
 		$params = array_merge( $params, $opt_params );
-		return $this->call( 'createAccountTicket', array( $params ), 'Google_Service_Analytics_AccountTicket' );
+		return $this->call( 'createAccountTicket', array( $params ), 'Google\Site_Kit_Dependencies\Google_Service_Analytics_AccountTicket' );
```
Fixes a PHP fatal: we need to pass the full class path/name here.

aaemnnosttv commented on assets/js/modules/analytics/common/account-field.js:
```diff
@@ -0,0 +1,41 @@
+ */
+import CreateAccountField from './create-account-field';
+
+const AccountField = ( { hasError, accountName, setAccountName } ) => {
+	return (
+		<CreateAccountField
+			hasError={ hasError }
+			value={ accountName }
+			setValue={ setAccountName }
```
Not sure there's much benefit in having field-specific props for each here.
* `accountName` could be `value`, and `setAccountName` could be `setValue`.
* `setValue` could be `onChange` and the API between components is more direct and clear.

Then all props could be passed through and you have less differences between each, and `CreateAccountField` is just a thin template component which is primarily concerned with DRYing up the markup for each.

felixarntz replied:
Agreed, I thought I had mentioned that before. Let's use `value` and `onChange` for all of these.

adamsilverstein replied:
Ah yes you did and I did part of this @felixarntz, will complete.

adamsilverstein replied:
Completed in https://github.com/google/site-kit-wp/pull/1389/commits/e97166b547e89d37cf74fe5c3d5f2b61ca6933fa

aaemnnosttv commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,133 @@
+						? <Select
+							className="googlesitekit-analytics__select-timezone"
+							name="timezone2"
+							style={ { minWidth: '240px' } /*todo: move to css */ }
```
Todo to do 😄

adamsilverstein replied:
Removed todo, and styles, already in css.

aaemnnosttv commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,133 @@
+		return response;
+	};
+
+	const timezoneSelector = useMemo( () => getTimezoneSelector(), [ selectedCountry, timezone, hasError ] );
```
Is this expensive to render? It seems unnecessary to memoize.

adamsilverstein replied:
Yes, it is expensive due to the large dataset and multiple level array. Without the memoization typing into other fields is noticeably slow

aaemnnosttv replied:
Hmm, I'll try to see if there are any easy perf-wins here but otherwise, let's at least add a comment as to why this is used for now 👍

adamsilverstein replied:
Added a comment.

aaemnnosttv replied:
See above - I don't think it's complex to render, but it's setting state above multiple times while iterating over the options which is likely the reason it is noticeably slower without the memo.

adamsilverstein replied:
see below: https://github.com/google/site-kit-wp/pull/1389#discussion_r420455380

aaemnnosttv commented on assets/js/modules/analytics/settings/settings-edit.js:
```diff
@@ -113,13 +114,15 @@ export default function SettingsEdit() {
 		};
 	}, [] );
 
+	const usingProxy = directSelect( 'core/site' ).isUsingProxy();
```
This should be using `useSelect`. We should never call `Data.select` directly from within a component.

adamsilverstein replied:
Since this never changes, can I use select outside the component? If I use `useSelect` what would the dependency be?

aaemnnosttv replied:
We shouldn't assume that data from the datastore will never change. Even if we don't explicitly change it via an action, all selectors will return `undefined` until resolved which is still a change to the return value that the component needs to be subscribed to. In most cases with simple data like this, the data will be preloaded so the resolution should be instant, but we shouldn't assume that. It's okay to use `useSelect` without any dependencies as this is different than an empty array which means "only run once".

adamsilverstein replied:
Ok, got it! thanks for clarifying, I will adjust these and remove the use of the direct `select` approach.

adamsilverstein replied:
Removed all direct `select` usage in favor of `useSelect`

aaemnnosttv commented on assets/js/modules/analytics/setup/setup-main.js:
```diff
@@ -69,6 +70,8 @@ export default function SetupMain( { finishSetup } ) {
 		trackEvent( 'analytics_setup', 'configure_analytics_screen' );
 	}, [] );
 
+	const usingProxy = directSelect( 'core/site' ).isUsingProxy();
```
See above 😄

adamsilverstein replied:
Fixed.

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+const { useDispatch, useSelect, select: directSelect } = Data;
+
+const AccountCreate = () => {
+	const { siteName, siteURL } = global.googlesitekit.admin;
```
I'd say we should add it to the `core/site` store as part of this issue like we did with `isUsingProxy`.

adamsilverstein replied:
Sure, I can add that. @aaemnnosttv then would we use `useSelect` here as well or can we use `select` and move this outside the component function?

aaemnnosttv replied:
> @aaemnnosttv then would we use useSelect here as well or can we use select and move this outside the component function?

Yes, we should always use `useSelect` so that the component updates properly if the value isn't immediately available. In this case, it should be, but we should avoid assuming that.

E.g. https://github.com/google/site-kit-wp/blob/f9d9806a6dc860dd64a5672ebcc802c584479374/assets/js/modules/analytics/common/anonymize-ip-switch.js#L40

Just to clarify, we should never use `Data.select` directly, even though it will be the same as what is passed to `useSelect` in production. One aspect is related to how hooks work and making sure the component updates with the datastore, but the other is related to tests. In tests, we create a fresh registry and register the stores on it all the time so in this case `registry != Data`. If you need access to the registry specifically, there is a `useRegistry` hook for inside a component, otherwise from within the datastore we always use `Data.commonActions.getRegistry` which yields the current registry (same with `createRegistrySelector` and `createRegistryControl`).

adamsilverstein replied:
Thanks for clarifying, now using `useSelect` everywhere and will here as well.

adamsilverstein replied:
added and used blog name from state in https://github.com/google/site-kit-wp/pull/1389/commits/83ea845f8ea1ea7a9b31e1959181934bdbb1aa17

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+		( select ) => {
+			return select( STORE_NAME ).getAccountTicketTermsOfServiceURL();
+		},
+		[]
```
Agreed - what's the reason for adding it here?

adamsilverstein replied:
I'm not sure actually, I must have copied this from somewhere. I'll add a dependency here...

adamsilverstein replied:
Added `isDoingCreateAccount` as a dependency. Not sure what to add for `isDoingCreateAccount`

adamsilverstein replied:
removed empty array there.

adamsilverstein commented on assets/js/googlesitekit/datastore/site/info.js:
```diff
@@ -278,6 +286,34 @@ export const selectors = {
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state Data store's state.
+	 * @return {?string} The timezone.
```
Will add.

adamsilverstein commented on assets/js/googlesitekit/datastore/site/info.js:
```diff
@@ -278,6 +286,34 @@ export const selectors = {
+	 * @return {?string} `true` if the proxy service is in use, `false` if not. Returns `undefined` if not loaded.
+	 */
+	isUsingProxy: createRegistrySelector( ( select ) => () => {
+		const { usingProxy } = select( STORE_NAME ).getSiteInfo() || false;
```
Fixed in https://github.com/google/site-kit-wp/pull/1389/commits/70da415139d2444a0cd40d2b9e1842c273da2ae3

adamsilverstein commented on assets/js/googlesitekit/datastore/site/info.js:
```diff
@@ -278,6 +286,34 @@ export const selectors = {
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state Data store's state.
+	 * @return {?string} `true` if the proxy service is in use, `false` if not. Returns `undefined` if not loaded.
```
? same as ?

aaemnnosttv replied:
Sorry, I was referring to the nullable `@return` that I thought I had flagged above on `getTimezone`.

Apologies for the confusion!

```suggestion
	 * @return {(string|undefined)} `true` if the proxy service is in use, `false` if not. Returns `undefined` if not loaded.
```

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+
+const AccountCreate = () => {
+	const { siteName, siteURL } = global.googlesitekit.admin;
+	let tz = directSelect( 'core/site' ).getTimezone();
```
@aaemnnosttv This is really only needed to grab the default timezone as passed from WordPress. I can switch to useSelect, however the value will never change so is that excessive? Can I move this up so it only runs once?

aaemnnosttv replied:
It's okay if it runs more than once (it's basically just returning the property of an object - state); the resolver is the expensive component to a selector and the datastore makes sure that it is only run once for each set of arguments.

adamsilverstein replied:
👍 updated.

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+			setIsNavigating( false );
+		}
+		send();
+	} );
```
Added

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+
+			// Redirect if the accountTicketTermsOfServiceURL is set.
+			if ( accountTicketTermsOfServiceURL ) {
+				location = accountTicketTermsOfServiceURL;
```
changed in https://github.com/google/site-kit-wp/pull/1389/commits/47c5c47733d1fc4473c5a3eab732afdc456950b5

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+	const handleSubmit = useCallback( ( accountName, propertyName, profileName, timezone ) => {
+		trackEvent( 'analytics_setup', 'new_account_setup_clicked' );
+		setIsNavigating( true );
+		async function send() {
```
I will give that another try, can't recall why I wound up doing it this way.

adamsilverstein replied:
Updated in https://github.com/google/site-kit-wp/pull/1389/commits/b3f703a13e86cc1d58194b228d3a325a6a483b16

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+	} );
+
+	// Fall back to the browser timezone if the WordPress timezone was not set.
+	if ( ! tz || '' === tz || 'UTC' === tz ) {
```
I will remove the special handling for UTC. I was under the impression this wasn't valid :) I'll also remove the browser timezone fallback since not the timezone will always be 'set'

adamsilverstein replied:
i kept the fallback for the case where we can't locate a matching timezone in the dropdown (as is the case with UTC, sorry WordPress).

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+			accountName: accountName === '',
+			propertyName: propertyName === '',
+			profileName: profileName === '',
+			timezone: timezone === '' || timezone === 'UTC', //An unset timezone in WordPress is reported as "UTC".
```
Yea, good point. Fixing.

adamsilverstein replied:
One issue with UTC: it isn't in our country data set as a timeZone->TimezoneID value (nor are any of the UTC+X timezones WordPress offers). In this case We can probably fall back to the browser timezone as a good first guess and let the user adjust if needed.

aaemnnosttv replied:
I misspoke 😄 (TIL) UTC is not a timezone, but a time standard which timezones are defined relative to!  The equivalent timezone is GMT, so we should map `UTC` to `Etc/GMT` given the list of IDs in `countries.js`.

adamsilverstein replied:
Ah, I wasn't sure about this. WordPress offers a bit of a weird set of choices. The dropdown in analytics is based on country. WordPress offers an offset based selector as well as countries. We don't need to try too hard here, most users will likely have selected a country which will match. If they have a UTC+ type timeszone, we will fall back to the browser as a guess.

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+		timezone: timezone === '',
+	} );
+
+	const validationHasIssues = Object.values( validationIssues ).some( ( check ) => check );
```
Nice!

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+		( select ) => {
+			return select( STORE_NAME ).isDoingCreateAccount();
+		},
+		[]
```
What would you suggest as a dependency for this `useSelect`?

adamsilverstein replied:
removed this one.

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	// Disable the submit button if there are validation errors, and while submission is in progress.
+	const buttonDisabled = validationHasIssues;
```
Switched to using `disabled={ validationHasIssues }`

aaemnnosttv commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,172 @@
+						</div>
+						<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-6">
+							<Button
+								disabled={ buttonDisabled }
```
```suggestion
								disabled={ validationHasIssues }
```

adamsilverstein replied:
already changed this, hadn't pushed :)

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,171 @@
+		tz = Intl.DateTimeFormat().resolvedOptions().timeZone;
+	}
+
+	const [ accountName, setAccountName ] = useState( siteName );
+	const [ propertyName, setPropertyName ] = useState( url.hostname );
+	const [ profileName, setProfileName ] = useState( __( 'All website traffic', 'google-site-kit' ) );
+	const [ timezone, setTimezone ] = useState( tz );
```
nothing bad happens in my testing :)

adamsilverstein commented:
@aaemnnosttv Thanks for all the feedback! I believe I addressed everything you raised, can you please give this another review?

aaemnnosttv commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,142 @@
+					>
+						{
+							allCountries && allCountries
+								.map( ( aCountry ) => {
+									// If the selected timezone is in this country, the country should be selected.
+									let value = aCountry.defaultTimeZoneId;
+									const timezoneMatch = aCountry.timeZone.find( ( tz ) => tz.timeZoneId === timezone );
+									if ( timezoneMatch ) {
+										foundTimezone = true;
+										value = timezoneMatch.timeZoneId;
+										if ( aCountry.timeZone.length > 1 ) {
+											multiTimezone = aCountry.timeZone;
+										} else {
+											setSelectedTimezoneID( aCountry.timeZone[ 0 ].displayName );
+											selectedTimezoneDisplay = aCountry.timeZone[ 0 ].displayName;
+											multiTimezone = false;
+										}
+										setSelectedCountry( timezoneMatch.timeZoneId );
+									}
+
+									return (
+										<Option
+											key={ aCountry.displayName }
+											value={ value }
+										>
+											{ aCountry.displayName }
+										</Option>
+									);
```
This function should only map countries into `Option`s, but there are `set` calls being made in the iteration.

This is likely rather unperformant (hence the need for `useMemo`) as you're setting state multiple times, every time this is rendered.

The `set*` calls should only be invoked by user-action (i.e. `onEnhancedChange`) or `useEffect` with specific dependencies as well as a guard - likely only to set if not already set.

adamsilverstein replied:
Good point. The set calls are about setting the selection status when we find a matching timezone in the data. I will break that part into a separate function and only call it once on initialization.

adamsilverstein replied:
@aaemnnosttv I tried moving the search out of the display logic, and this does remove the set state calls, however I found I still need the caching to make the dropdown performant. See https://github.com/google/site-kit-wp/pull/1389/commits/d639a51a9b3f2cada4f255ddfc4b0a6e4d18e002 and let me know what you think.

Part of the issue here is that this is a two part dropdown, the second part depends on the first part (when you change the country, you may or may not get a second dropdown with timezones for that country).

Adding useMemo ensures the the drop-downs only re-render when timezone changes and seems reasonable, however I am open to any optimizations you can find please feel free to push updates if you have ideas.

adamsilverstein replied:
I managed to simplify a bit further by only calling `timezoneInCountries` once when we first load the component. I run this once using a flag `timezoneChecked` which is set to true after the first run. The main point of this function is to see if the timezone set in WordPress can be matched to an actual timezone and to fall back to the browser timezone if not.

Documented better in https://github.com/google/site-kit-wp/pull/1389/commits/552bd9cf0174d32d7f4c3fe0a9795dc065a83012.

aaemnnosttv commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,142 @@
+							{
+								multiTimezone
+									.map( ( aTimezone ) => {
+										if ( aTimezone.timeZoneId === timezone ) {
+											foundTimezone = true;
+										}
```
Same here - albeit primitive, there shouldn't be any non-mapping logic in here.

Also, the same list of `<Options>` should be reusable here rather than generating it again.

adamsilverstein replied:
Right, I'll remove this part as well. We should only need to regenerate if the user chooses another country.

adamsilverstein replied:
removed this

aaemnnosttv commented on assets/js/modules/analytics/settings/settings-edit.js:
```diff
@@ -47,6 +48,7 @@ export default function SettingsEdit() {
 	const isDoingSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).isDoingSubmitChanges() );
 	const haveSettingsChanged = useSelect( ( select ) => select( STORE_NAME ).haveSettingsChanged() );
 	const isCreateAccount = ACCOUNT_CREATE === accountID;
+	const usingProxy = useSelect( ( select ) => select( 'core/site' ).isUsingProxy() );
```
This should be importing `STORE_NAME` from the core site store rather than hard coding it.
Otherwise, this is beautiful!

adamsilverstein replied:
👍 Will do!

adamsilverstein replied:
Done.

aaemnnosttv commented on assets/js/modules/analytics/setup/setup-main.js:
```diff
@@ -47,6 +48,7 @@ export default function SetupMain( { finishSetup } ) {
 	const isDoingSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).isDoingSubmitChanges() );
 	const hasResolvedAccounts = useSelect( ( select ) => select( STORE_NAME ).hasFinishedResolution( 'getAccounts' ) );
 	const isCreateAccount = ACCOUNT_CREATE === accountID;
+	const usingProxy = useSelect( ( select ) => select( 'core/site' ).isUsingProxy() );
```
Same as above regarding `STORE_NAME`.

adamsilverstein replied:
👍

aaemnnosttv replied:
This one still needs to be updated.

adamsilverstein commented on assets/js/modules/analytics/common/profile-field.js:
```diff
@@ -0,0 +1,41 @@
+			hasError={ hasError }
+			value={ value }
+			setValue={ setValue }
+			label={ __( 'View', 'google-site-kit' ) } // Profiles are now called views.
```
👍 That change was based on direct feedback. I will revert for now and we can rename along with everything else.

adamsilverstein replied:
Looks like we already call this "View" in our existing "profile" selector? https://github.com/google/site-kit-wp/blob/253514fbd49227920b6818125fd9a05e168c2ead/assets/js/modules/analytics/common/profile-select.js#L61

aaemnnosttv replied:
You're right, I didn't realize that. We do however say "Set up a new profile" which is confusing.

This should definitely stay "View" then, but there are other occurrences to fix in 1486.

adamsilverstein commented:
>  I don't think it's complex to render,
Yea, mostly the complexity is tracking down the timezone that matches the user. I will refactor this to separate it out and only run on initialization. Thanks for the review.

adamsilverstein commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,142 @@
+						outlined
+					>
+						{
+							allCountries && allCountries
```
nope, removing.

adamsilverstein replied:
removed

adamsilverstein commented on assets/js/modules/analytics/common/timzezone-select.js:
```diff
@@ -0,0 +1,142 @@
+						enhanced
+						value={ selectedCountry }
+						onEnhancedChange={ ( i, item ) => {
+							setTimezone( item.dataset.value );
+							setSelectedCountry( item.dataset.value );
+							setSelectedTimezoneID( item.dataset.value );
```
So this would be 
```
setTimezone( item.dataset );
setSelectedCountry( item.dataset );
setSelectedTimezoneID( item.dataset );
```
?

adamsilverstein replied:
@aaemnnosttv ^^

adamsilverstein commented:
@aaemnnosttv I hope I addressed all your issues. Github keeps hiding comments so I may have missed something. I am going to click 'resolve' for everything I know I addressed.

adamsilverstein commented:
@aaemnnosttv A few items to review here when you have a moment, I still haven't managed to remove the need to memoize the timezone selector.

aaemnnosttv commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,180 @@
+			timezone,
+		} );
+
+		if ( result.error ) {
+			setIsNavigating( false ); // Silently fail for server errors.
+		}
```
This can be simplified a bit as the progress bar is already shown if doing create account. `setIsNavigating( true )` is only needed if create account was successful.

adamsilverstein replied:
simplified in https://github.com/google/site-kit-wp/pull/1389/commits/1959e34b9e88cfe463f38790f7de7899b6ac9c76

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,180 @@
+	const url = new URL( siteURL );
+	const { createAccount } = useDispatch( STORE_NAME );
+
+	const isDoingCreateAccount = useSelect( ( select ) => select( STORE_NAME ).isDoingCreateAccount() );
+
+	const accountTicketTermsOfServiceURL = useSelect( ( select ) => select( STORE_NAME ).getAccountTicketTermsOfServiceURL() );
```
Right, that makes sense, I'll move everything up as much as possible.

adamsilverstein replied:
Moved in https://github.com/google/site-kit-wp/pull/1389/commits/c3926c816a823fd3099377358ca1c032101a04fe.

adamsilverstein commented:
> The error is set in state by CATCH_FETCH_CREATE_ACCOUNT and displayed automatically in ErrorNotice if present.

Added in https://github.com/google/site-kit-wp/pull/1389/commits/15c510b83789fe4726c1ed319312d719381f8a0a.

This almost worked! The component did catch the error, however the message is my test case where I'm triggering a fatal in PHP has no message:

![image](https://user-images.githubusercontent.com/2676022/81184024-21d47380-8f6d-11ea-9e52-1791ca8a89bb.png)

I assume this would work better for actual errors, still I wonder: should there be a fallback somewhere to enable errors with no message ("Unknown error")?

aaemnnosttv commented:
@googlebot I consent.

adamsilverstein commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -85,6 +72,23 @@ export default function AccountCreate() {
 	}, [ timezone ] );
 
+	const handleSubmit = useCallback(
+		async () => {
```
Nice!

aaemnnosttv commented:
@adamsilverstein please hold off on adding any more commits here for now. After digging into this with @tofumatt a bit more we found that there were still substantial performance issues which seem to be related to the centralized state in `AccountCreate` causing all the components to re-render when one of them changes. Preliminary testing using a simple datastore-based approach resolves this so I will be pushing a few more changes to replace (most) of the local state usage 😄

adamsilverstein commented:
@aaemnnosttv Testing the latest version, everything looks good. Testing in the debugger I see that only edited components get updates - much better! Tying it all together with the datastore approach neatly eliminates the top level state tracking (validation, etc.) that was causing the updates.

Thanks for taking the time to dig into this!

tofumatt commented on assets/js/googlesitekit/datastore/site/info.js:
```diff
@@ -278,6 +290,48 @@ export const selectors = {
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state Data store's state.
+	 * @return {(string|undefined)} `true` if the proxy service is in use, `false` if not. Returns `undefined` if not loaded.
```
```suggestion
	 * @return {(boolean|undefined)} `true` if the proxy service is in use, `false` if not. Returns `undefined` if not loaded.
```

tofumatt commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,131 @@
+	const isDoingCreateAccount = useSelect( ( select ) => select( STORE_NAME ).isDoingCreateAccount() );
+	const siteURL = useSelect( ( select ) => select( CORE_SITE ).getReferenceSiteURL() );
+	const siteName = useSelect( ( select ) => select( CORE_SITE ).getSiteName() );
+	const tz = useSelect( ( select ) => select( CORE_SITE ).getTimezone() );
```
Can we can this `timezone` instead? `tz` is a bit vague and tough to scan. We can spare the bytes 😉 
```suggestion
	const timezone = useSelect( ( select ) => select( CORE_SITE ).getTimezone() );
```

aaemnnosttv replied:
They were named differently before because this was being fed into `useState` as the default value and the state variable was `timezone`. We kind of have a similar thing now where this selector gets the timezone from WP whereas we need `timezone` from the form.

adamsilverstein replied:
using timezone now.

tofumatt commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,131 @@
+	const { setForm } = useDispatch( STORE_NAME );
+	useEffect( () => {
+		const { hostname } = new URL( siteURL );
+		let timezone = tz;
```
Argh, is this why the above variable is named as such? I dunno, this seems like it could be simplified. Maybe remove this line and see my ternary suggestion below?

adamsilverstein replied:
removed `tz`!

tofumatt commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,131 @@
+			propertyName: hostname,
+			profileName: __( 'All website traffic', 'google-site-kit' ),
+			countryCode: countryCodesByTimezone[ timezone ],
+			timezone,
```
Usually not a fan of ternaries, but in this case it allows us to avoid the two `timezone` variables.
```suggestion
			timezone: ! countryCodesByTimezone[ timezone ] ? Intl.DateTimeFormat().resolvedOptions().timeZone : timezone,
```

(assuming you rename `tz` to `timezone`)

adamsilverstein replied:
refactored to avoid tz usage in https://github.com/google/site-kit-wp/pull/1389/commits/e1deb02800ce9ed14beed8ff4ec4a80656e6547e. We have to actually set timezone to the browser value as well. You can test this by setting WordPress timezone to "UTC" (or any of the UTC+ options). This is not a "real" timezone, so we fall back to the browser timezone.

tofumatt commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,131 @@
+				{ __( 'Create new Analytics account', 'google-site-kit' ) }
+			</h3>
+
+			{ __( 'We’ve pre-filled the required information for your new account. Confirm or edit any details:', 'google-site-kit' ) }
```
This should be in some kind of markup, like a `<p>` or something, no?
```suggestion
			<p>{ __( 'We’ve pre-filled the required information for your new account. Confirm or edit any details:', 'google-site-kit' ) }</p>
```

adamsilverstein replied:
👍

tofumatt commented on assets/js/modules/analytics/common/account-create/create-account-field.js:
```diff
@@ -0,0 +1,55 @@
+			) }
+			label={ label }
+			name={ name }
+			onChange={ ( e ) => {
```
`event` is more readable 😉

adamsilverstein replied:
updated in https://github.com/google/site-kit-wp/pull/1389/commits/01b7581484ef5ec3e020a4da099e6407b5744a76

tofumatt commented on assets/js/modules/analytics/common/account-create/profile-field.js:
```diff
@@ -0,0 +1,51 @@
+
+	return (
+		<CreateAccountField
+			label={ __( 'View', 'google-site-kit' ) } // Profiles are now called views.
```
Thanks for the comment, that's super-helpful 😄

tofumatt commented on assets/js/modules/analytics/datastore/form.js:
```diff
@@ -0,0 +1,76 @@
+/**
```
This is great, I dig it! 👍

tofumatt commented on assets/js/modules/analytics/datastore/form.js:
```diff
@@ -0,0 +1,76 @@
+	switch ( type ) {
+		case SET_FORM:
+			const { formID, data } = payload;
+			const formKey = `form::${ formID }`;
```
Any reason for the key prefix?

aaemnnosttv replied:
So a `formID` can't conflict with other state in the store. We could put that in the formID itself, but it seemed a bit safer this way.

tofumatt commented on assets/js/modules/analytics/util/countries-data.js:
```diff
@@ -0,0 +1,3042 @@
+export const countries = {
```
This feels like a precarious thing for us to maintain—is there an npm library that'd get updates with the same type of data?

aaemnnosttv replied:
I had the same thought. This is essentially a dump from an Analytics API endpoint so it's guaranteed to match. I think we should double check to see if this can be replaced with a module though.

felixarntz replied:
Agreed. Maybe we can even request it on demand if it's not available in a module. I'm just wary about this endpoint unexpectedly changing, since it's apparently not documented. Let's follow up on this @adamsilverstein

felixarntz commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,127 @@
+			accountName: siteName,
+			propertyName: hostname,
+			profileName: __( 'All website traffic', 'google-site-kit' ),
+			countryCode: countryCodesByTimezone[ timezone ],
```
This can potentially still result in `countryCode` being `undefined`, since we don't 100% know if there's a match for every browser timezone. Looking at the rest of the code though, this should be fine and it would just not select an initial value in the dropdown (which is okay because it's likely an edge-case, and we'd simply just not know the right default).

felixarntz commented on assets/js/modules/analytics/common/account-create/country-select.js:
```diff
@@ -0,0 +1,73 @@
+		if ( newCountryCode !== value ) {
+			setForm( FORM_ACCOUNT_CREATE, {
+				countryCode: newCountryCode,
+				timezone: countriesByCode[ newCountryCode ].defaultTimeZoneId,
```
This is _probably_ not going to happen since the potential values are sourced by the same data, but it may be a bit safer to add a check for `countriesByCode[ newCountryCode ]` in the if clause above. If it were to be `undefined` here, we'd get a JavaScript error.

But again, given the data this should never happen. I'd rather do defensive coding regardless.

aaemnnosttv commented on assets/js/modules/analytics/datastore/form.js:
```diff
@@ -0,0 +1,76 @@
+
+export const selectors = {
+	getForm( state, formID, key ) {
+		const form = state[ `form::${ formID }` ] || {};
```
That was how I had it originally but it made for a rather verbose reducer hence the compound key. I wouldn't be opposed to doing that though if we had a clean way to set nested data in an object. We should be able to use something like [`_.set()`](https://lodash.com/docs/4.17.15#set) for this though.

felixarntz replied:
Let's not worry about this here, but consider it for the proper `core/forms` store.

aaemnnosttv commented on assets/js/modules/analytics/common/account-create.js:
```diff
@@ -0,0 +1,127 @@
+	return (
+		<div>
+			<h3 className="googlesitekit-heading-4">
+				{ __( 'Create new Analytics account', 'google-site-kit' ) }
```
@felixarntz shouldn't this be "Create _a_ new Analytics Account" ?

```suggestion
				{ __( 'Create a New Analytics Account', 'google-site-kit' ) }
```

felixarntz replied:
Not sure about capitalizing these words since they're not part of a name or anything. @adamsilverstein This copy was reviewed by @marrrmarrr already right?

aaemnnosttv replied:
@felixarntz The case change here would only be if it's the title of the view. If it's not really a title, and just a heading, then the original casing is fine. I'm actually more concerned about the missing "a" there 😄

felixarntz replied:
Per follow-up convo with @marrrmarrr changing this to "Create your Analytics account".
